common/searchattribute/name_type_map.go

```diff
@@ -117,3 +118,13 @@ func (m NameTypeMap) IsDefined(name string) bool {
 	}
 	return false
 }
+
+func (m NameTypeMap) Copy() NameTypeMap {
+	return NameTypeMap{
+		customSearchAttributes: maps.Clone(m.customSearchAttributes),
+	}
+}
+
+func (m NameTypeMap) AddCustomSearchAttribute(name string, value enumspb.IndexedValueType) {
+	m.customSearchAttributes[name] = value
+}
```

Comment:
This is an area where my Go knowledge is limited. Without exposing a mutation and without exporting `customSearchAttributes`, how can a new attribute be added? The only way I know is to modify Copy:

```
func (m NameTypeMap) Copy(additionalAttributes map[string]enumspb.IndexedValueType) NameTypeMap {
    newMap := NameTypeMap{
        customSearchAttributes: maps.Clone(m.customSearchAttributes),
    }
    
    // Add the additional attributes
    for key, value := range additionalAttributes {
        newMap.customSearchAttributes[key] = value
    }
    
    return newMap
}
```

Reply:
My intention was to avoid having multiple maps that are very similar and instead add what is needed when it is needed, but if the map duplication is preferred, that's fine, too.